astacus/coordinator/plugins/clickhouse/steps.py

```diff
@@ -811,8 +821,46 @@ async def run_step(self, cluster: Cluster, context: StepsContext) -> None:
                 await disk_object_storage.delete_item(key_to_remove)
 
 
+@dataclasses.dataclass
+class GetVersionsStep(Step[List[ClickHouseVersion]]):
+    """Get the version of ClickHouse running on every node."""
+
+    clients: Sequence[ClickHouseClient]
+
+    async def run_step(self, cluster: Cluster, context: StepsContext) -> List[ClickHouseVersion]:
+        return await asyncio.gather(*(get_version(client) for client in self.clients))
+
+
+async def get_version(clickhouse_client: ClickHouseClient) -> ClickHouseVersion:
+    rows = await clickhouse_client.execute(b"SELECT version()")
+    assert isinstance(rows[0][0], str)
+    major, minor = rows[0][0].split(".")[:2]
+    return int(major), int(minor)
+
+
 async def execute_with_timeout(client: ClickHouseClient, timeout: float, query: bytes) -> None:
     # we use a session because we can't use the SETTINGS clause with all types of queries
     session_id = secrets.token_hex()
     await client.execute(f"SET receive_timeout={timeout}".encode(), session_id=session_id)
     await client.execute(query, session_id=session_id, timeout=timeout)
+
+
+CLICKHOUSE_VERSION_PARTITION_CMDS_NOT_REPLICATED = (23, 8)
```

Comment:
This seem oddly named, what we're really after is the version where ALTER ... FREEZE became unreplicated, not ALTER ... PARTITION.

Reply:
I guess I was using the same language as here https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/PartitionCommands.h#L19.  Previously all of these were replicated except ATTACH, FETCH and DROP.  Now none are replicated.

This is a bit obscure and no one will understand this so I will change to specifically refer to FREEZE.

Reply:
Makes sense :+1: